src/plugins/data_source_management/public/components/data_source_view/data_source_view.tsx

```diff
@@ -57,77 +63,140 @@ export class DataSourceView extends React.Component<DataSourceViewProps, DataSou
       showEmptyState: false,
       showError: false,
       defaultDataSource: null,
+      defaultDataSourceOption: null,
     };
   }
 
   componentWillUnmount() {
     this._isMounted = false;
   }
+  updateSelectedOptionState(defaultDataSource: string | null, dataSourceOption: DataSourceOption) {
+    this.setState({
+      selectedOption: [dataSourceOption],
+      defaultDataSource,
+    });
+    this.props.onSelectedDataSources?.([dataSourceOption]);
+  }
+
+  clearSelectedOptionState() {
+    this.setState({ selectedOption: [] });
+    this.props.onSelectedDataSources?.([]);
+  }
+
+  async getAndSetDefaultDataSourceOption(dataSourceId: string, defaultDataSource: string | null) {
+    try {
+      const defaultDataSourceObj = defaultDataSource
+        ? await this.getDefaultDataSourceObj(defaultDataSource)
+        : null;
+      const filteredDefaultDataSourceOption = this.getFilteredDataSource(defaultDataSourceObj);
+      this.setState({ defaultDataSourceOption: filteredDefaultDataSourceOption });
+    } catch (error) {
+      this.handleInvalidDataSourceError(dataSourceId);
+    }
+  }
+
+  async getSelectedDataSource(dataSourceId: string, defaultDataSource: string | null) {
+    try {
+      return await getDataSourceById(dataSourceId, this.props.savedObjectsClient!);
+    } catch (error) {
+      await this.getAndSetDefaultDataSourceOption(dataSourceId, defaultDataSource);
+      this.handleInvalidDataSourceError(dataSourceId);
+    }
+  }
+
+  handleInvalidDataSourceError(dataSourceId: string) {
+    const handleDataSourceViewErrorParams = this.getDataSourceViewErrorParams(
+      dataSourceId,
+      this.state.defaultDataSourceOption
+    );
+    handleDataSourceViewErrorWithSwitchToDefaultOption(handleDataSourceViewErrorParams);
+  }
+
+  async handleDataSourceLabelOrBackupWithDefaultOption(
+    dataSourceId: string,
+    defaultDataSource: string | null
+  ) {
+    const selectedDataSource = await this.getSelectedDataSource(dataSourceId, defaultDataSource);
+    if (!this._isMounted) return;
+    const filteredSelectedDataSourceOption = this.getFilteredDataSource(selectedDataSource);
+    // if the selectedOption has been filtered out, treat it as invalid id error and early return
+    if (filteredSelectedDataSourceOption) {
+      this.setState({
+        selectedOption: [filteredSelectedDataSourceOption],
+        defaultDataSource,
+      });
+      return;
+    }
+
+    await this.getAndSetDefaultDataSourceOption(dataSourceId, defaultDataSource);
+    this.handleInvalidDataSourceError(dataSourceId);
```

Comment:
are we calling this.handleInvalidDataSourceError twice if there is exception in this.getAndSetDefaultDataSourceOption

Reply:
Woops, it did call twice then, let me update the getAndSetDefaultDataSourceOption to only return error not call handleInvalidDataSourceError